src/wallet/wallet.cpp

```diff
@@ -1639,10 +1640,25 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate, b
             ReadBlockFromDisk(block, pindex);
             int posInBlock;
             for (posInBlock = 0; posInBlock < (int)block.vtx.size(); posInBlock++) {
-                if (AddToWalletIfInvolvingMe(*block.vtx[posInBlock], pindex, posInBlock, fUpdate))
+                const auto& tx = block.vtx[posInBlock];
+                if (AddToWalletIfInvolvingMe(*tx, pindex, posInBlock, fUpdate)) {
+                    myTxHashes.push_back(tx->GetHash());
                     ret++;
+                }
             }
 
+            // Sapling
+            SaplingMerkleTree saplingTree;
+            // This should never fail: we should always be able to get the tree
+            // state on the path to the tip of our chain
+            if (pindex->pprev) {
+                if (Params().GetConsensus().NetworkUpgradeActive(pindex->pprev->nHeight,  Consensus::UPGRADE_V5_DUMMY)) {
+                    assert(pcoinsTip->GetSaplingAnchorAt(pindex->pprev->hashFinalSaplingRoot, saplingTree));
+                }
+            }
+            // Increment note witness caches
+            ChainTipAdded(pindex, &block, saplingTree);
```

Comment:
(ee3f99a3d8eed08e48dcfd22356889f0d0117bdf)
we can probably put this inside the previous nested `if`s (skipping it before Sapling activation).